sparkmagic/sparkmagic/livyclientlib/sparkcommand.py

```diff
@@ -0,0 +1,147 @@
+        else:
+            command = u'{}.collect()'.format(command)
+        # Unicode support has improved in Python 3 so we don't need to encode.
+        if encode_result:
+            print_command = '{}.encode("{}")'.format(constants.LONG_RANDOM_VARIABLE_NAME,
+                                                     conf.pyspark_python_encoding())
+        else:
+            print_command = constants.LONG_RANDOM_VARIABLE_NAME
+        command = u'for {} in {}: print({})'.format(constants.LONG_RANDOM_VARIABLE_NAME,
+                                                    command,
+                                                    print_command)
+        return Command(command)
+
+    def _scala_command(self, spark_context_variable_name):
+        command = u'{}.toJSON'.format(spark_context_variable_name)
+        if self.samplemethod == u'sample':
+            command = u'{}.sample(false, {})'.format(command, self.samplefraction)
+        if self.maxrows >= 0:
+            command = u'{}.take({})'.format(command, self.maxrows)
+        else:
+            command = u'{}.collect'.format(command)
+        return Command(u'{}.foreach(println)'.format(command))
+
+    def _r_command(self, spark_context_variable_name):
+        command = spark_context_variable_name
+        if self.samplemethod == u'sample':
+            command = u'sample({}, FALSE, {})'.format(command,
+                                                      self.samplefraction)
+        if self.maxrows >= 0:
+            command = u'take({},{})'.format(command, self.maxrows)
+        else:
+            command = u'collect({})'.format(command)
+        command = u'jsonlite::toJSON({})'.format(command)
+        command = u'for ({} in ({})) {{cat({})}}'.format(constants.LONG_RANDOM_VARIABLE_NAME,
+                                                         command,
+                                                         constants.LONG_RANDOM_VARIABLE_NAME)
+        return Command(command)
+
+
+    @staticmethod
+    def _records_to_dataframe(records_text, kind):
+        if records_text in ['', '[]']:
+            strings = []
+        else:
+            strings = records_text.split('\n')
+        try:
+            data_array = [json.JSONDecoder(object_pairs_hook=OrderedDict).decode(s) for s in strings]
+
+            if kind == constants.SESSION_KIND_SPARKR and len(data_array) > 0:
+                data_array = data_array[0]
+
+            if len(data_array) > 0:
+                df = pd.DataFrame(data_array, columns=data_array[0].keys())
+            else:
+                df = pd.DataFrame(data_array)
+
+            coerce_pandas_df_to_numeric_datetime(df)
+            return df
+        except ValueError:
+            raise DataFrameParseException(u"Cannot parse object as JSON: '{}'".format(strings))
+
+    # Used only for unit testing
+    def __eq__(self, other):
+        return self.code == other.code and \
+            hasattr(other, 'samplemethod') and \
```

Comment:
SparkStoreCommand inherits from Command. assert_any_call and similar functions throw an exception when comparing instances of the two. Haven't checked if this is bug in nose or expected behavior.